### PR TITLE
Use require.resolve as a fallback of path.resolve

### PR DESCRIPTION
**Pull request by kachkaev**
Hi @DavidAnson! We had a [discussion](https://github.com/igorshubovych/markdownlint-cli/issues/97) about resolving `config.extends` in the CLI some time ago; this PR is a follow-up.

My initial question a few months ago was about extending `js` files in an auto-discovered `json` config and you explained some unwanted security implications of a change in the logic. We also discussed an opportunity of using `require.resolve` as a risk-free option for those who want to use a shared markdownlint config and avoid writing something like this in their `.markdownlint.json`:

```js
{
  "extends": "./node_modules/@my-company/markdownlint-config/index.json",
  "first-line-heading": false,
  "no-inline-html": false
}
```

This is what used to stick with till now.

This weekend [I was playing](https://github.com/kachkaev/njt/pull/29) with [Yarn v2 (Berry)](https://github.com/yarnpkg/berry) and realised that the conversion of `"./node_modules/@my-company/markdownlint-config/index.json"` to `"@my-company/markdownlint-config"` became unavoidable. The folder called `node_modules` no longer exists and the real location of all dependency files is now fully controlled by Yarn. Thus, there is no way to refer to a config file in a third-party package without using `require.resolve`.

This PR adds `require.resolve` as a fallback to `path.resolve` in a fully backwards-compatible manner. It should not affect the security model and does not change the error messages in the existing scenarios. The resolution strategy just becomes a little bit more optimistic and can successfully handle more cases than it used to before. So the change in the logic remains unnoticeable for the existing Markdownlint users, while making it possible to use fancier and shorter `"extends"` paths.

I'd be keen to hear what you think about this change and if you're generally happy with the approach, I'm happy to add tests and resolve any other review comments you might have. WDYT?

### CODE REVIEW

**Comment by DavidAnson** [comment]
Since then, I have used `require.resolve` in CLI2, so I am more familiar with it. Thanks for bringing this up again! I will have a look at the changes. I'd definitely want to add test cases for the new scenarios. FYI that I'm preparing for a patch release of the library shortly, so would wait to merge this PR until that went out.

File: lib/markdownlint.js
Comment (DavidAnson) [inline lib/markdownlint.js:959]: Can we call it extendsFile like before? Will give a smaller diff.

File: lib/markdownlint.js
Comment (DavidAnson) [inline lib/markdownlint.js:925]: Diff would be smaller without this change, too. I think before/after are pretty similar - let's limit this to what's needed?

File: lib/markdownlint.js
Comment (DavidAnson) [inline lib/markdownlint.js:973]: Same comments as other function.

File: lib/markdownlint.js
Comment (DavidAnson) [inline lib/markdownlint.js:899]: "ina"

File: lib/markdownlint.js
Comment (DavidAnson) [inline lib/markdownlint.js:903]: Add type, string, right?

File: lib/markdownlint.js
Comment (DavidAnson) [inline lib/markdownlint.js:904]: Add description.

File: lib/markdownlint.js
Comment (DavidAnson) [inline lib/markdownlint.js:913]: This is used by sync and async callers, right? Goal is to keep sync calls out of the async path.

File: lib/markdownlint.js
Comment (DavidAnson) [inline lib/markdownlint.js:906]: Can we call it "referenceId", maybe - it might not be a file anymore?

File: lib/markdownlint.js
Comment (DavidAnson) [inline lib/markdownlint.js:924]: Are these catches to be extra safe, or would both methods throw in normal use? I prefer to let true exceptions bubble out.

**Review by DavidAnson — CHANGES_REQUESTED** [review]
Makes sense! I left some comments/questions. Will do another review after you have a chance to go over them. Thanks!

**Comment by kachkaev** [comment]
Thanks for a quick reply @DavidAnson! I'm happy to address your comments within 12 hours – just going to sleep now 🙂

**Comment by kachkaev** [comment]
Alternatively, feel free to tweak the PR yourself you that’s preferred. I don’t worry much about the authorship, so will be also happy is you close this PR and implement require.resolve fallback yourself – whatever you find easier. Speak soon!

**Comment by DavidAnson** [comment]
No hurry. Get some sleep!

File: test/markdownlint-test.js
Comment (kachkaev) [inline test/markdownlint-test.js:1035]: I could not find alternatives to Jest’s `before()` / `after()` hooks, so followed the idea in https://github.com/substack/tape/issues/59#issuecomment-32809481

File: lib/markdownlint.js
Comment (kachkaev) [inline lib/markdownlint.js:922]: I had to add this line after facing quite a mystery. Without it, running `npm run ci; npx c8 report --reporter html` _on Node 12_ results in this:

Screenshot 2020-10-18 at 12 43 30

This is a false-positive because both branches are covered. If I add `console.log` in both, they do show up. Other versions though [don’t suffer from this problem](https://github.com/DavidAnson/markdownlint/runs/1270965501). Thoughts?

File: lib/markdownlint.js
Comment (kachkaev) [reply lib/markdownlint.js:911]: Generally speaking, I agree with the principle. We can duplicate `resolveConfigExtends` so that we have `resolveConfigExtends` and `resolveConfigExtendsSync` like everywhere else, but here are the things to consider:

- `fs.statSync` is super quick, it’s not the same as calling `fs.readFileSync` or `fs.writeFileSync`, which take forever (on CPU scale)
- Node’s `require.resolve` _is_ sync, which in turn depends on some _sync_ fs calls. There can be quite a few of them actually, which involve getting stats different paths, reading `package.json` file and so on.
- Splitting will add more lines to the file, which increases the surface of the bugs
- An async version of the function will make `readConfig` harder to read because we use callbacks and not async / await
- Splitting will require more tests to be added

Happy for you to make a call on this. If a sync nature of `require.resolve` → `fs` is not a sufficient argument for a trade-off I've made, happy to split the function.

File: lib/markdownlint.js
Comment (kachkaev) [reply lib/markdownlint.js:924]: The key principle I was trying to follow is backwards-compatibility, including the kinds of error messages people get for the unresolved paths. This is what resulted this logic.

If the original `path.resolve` could not throw, `resolveConfigExtends` should not either. It should first try to resolve the reference using `path.resolve` as before _and only if_ the resulting path does not point to an existing _file_ (not folder), try `require.resolve`. By default, `require.resolve` throws, so if we re-throw, the resulting error message ends up being different from previous versions. Thus, we’d have to qualify this PR as a breaking change, which I don’t think is necessary.

That said, I’m open to suggestions 🙂

File: lib/markdownlint.js
Comment (DavidAnson) [reply lib/markdownlint.js:911]: That's a good point, let's stick with this for now.

File: lib/markdownlint.js
Comment (DavidAnson) [reply lib/markdownlint.js:924]: I'm thinking of this as a feature and want it to go out as a semver-breaking change. So I'm not worried about the kind of exception changing. Everything that used to work should continue to behave the same - new scenarios can start working - but things that didn't work can not work differently now. :)

File: lib/markdownlint.js
Comment (DavidAnson) [reply lib/markdownlint.js:922]: C8 only seems to be reliable in the LATEST Node (14). In fact, it already does not run in the CI workflow for this project for Node 10. On another project of mine, it ONLY runs on Node 14. If the behavior you're seeing here is correct on Node 14, we can tweak `nodejs.yml` to do the same here.

File: test/config/pseudo-package/package.json
Comment (DavidAnson) [inline test/config/pseudo-package/package.json:2]: Let's just call the file `config-frompackage.json` or something like that.

File: test/config/config-parent-of-package.json
Comment (DavidAnson) [inline test/config/config-parent-of-package.json:1]: Maybe call this file `config-packageparent.json`?

File: test/markdownlint-test.js
Comment (DavidAnson) [reply test/markdownlint-test.js:1035]: I'm not sure the symlink will work on Windows. And introducing an ordering dependency here will cause problems later when switching to an async runner. Can we move this to a copy+delete in the single test function? I know if it throws an exception it won't clean up all the way, but that's fine with me.

File: test/markdownlint-test.js
Comment (DavidAnson) [inline test/markdownlint-test.js:1040]: Could we place this into a `node_modules` under the `test` directory instead? That will keep it out of the "real" one and it should still be searched first, I think?

File: test/markdownlint-test.js
Comment (DavidAnson) [inline test/markdownlint-test.js:1110]: "file" -> "package"

Comment (DavidAnson) [reply test/markdownlint-test.js:1110]: For lines 1104/1108/1109

File: test/markdownlint-test.js
Comment (DavidAnson) [inline test/markdownlint-test.js:1107]: In this case of requiring a module, I think `MODULE_NOT_FOUND` is a better code (per discussion above).

**Review by DavidAnson — CHANGES_REQUESTED** [review]
This is looking really good, thank you!

File: lib/markdownlint.js
Comment (kachkaev) [reply lib/markdownlint.js:924]: IC. How would you implement resolution? There are several ways, each with its own trade-offs. So I'm curious to know your vision, including the priority of the resolutions and the expected feedback to users.

File: lib/markdownlint.js
Comment (merceyz) [inline lib/markdownlint.js:920]:
```suggestion
    return require.resolve(referenceId, { paths: [configFile] });
```

This must be done on behalf of the configFile, otherwise it will be required from `markdownlint` which doesn't declare it as a dependency. Under PnP this will throw an error and under the node_modules layout makes it rely on hoisting to bring `referenceId` and `markdownlint` to the same level.

File: lib/markdownlint.js
Comment (kachkaev) [reply lib/markdownlint.js:914]: `fs.statSync(resolvedExtendsFile)` would not throw for directories. Given that `require.resolve` comes second, we probably want to try it in all cases except when a file is found. The same problem is with `fs.exists` (it returns true for files _and_ folders).

File: test/markdownlint-test.js
Comment (kachkaev) [reply test/markdownlint-test.js:1110]: The problem is that this error comes from `fs.readFile` / `fs.readFileSync` in `readConfig`/`readConfigSync`. At that point we don't know if the resolution was done via `path.resolve` or `require.resolve`. Besides, if a config is like `@my-company/my-package/my-file.json` it may be that the package is resolved but the file in the package is non-existent.

File: test/markdownlint-test.js
Comment (kachkaev) [reply test/markdownlint-test.js:1107]: What about `@my-company/my-package/my-file.json` (assuming the package exists, but the file within it does not)?

File: lib/markdownlint.js
Comment (kachkaev) [reply lib/markdownlint.js:920]: Hi @merceyz! 👋

I tried adding that parameter in my earlier experiments, but this broke the currently tested behaviour. I need to double-check.

File: lib/markdownlint.js
Comment (kachkaev) [reply lib/markdownlint.js]: I tweaked `nodejs.yml` as you suggested, CI passed. Thanks! Resoving.

File: lib/markdownlint.js
Comment (kachkaev) [reply lib/markdownlint.js]: It worked! 👍

File: .github/workflows/nodejs.yml
Comment (fisker) [inline .github/workflows/nodejs.yml:33]:
```suggestion
      if: ${{ matrix.node-version != '14.x' }}
```

File: lib/markdownlint.js
Comment (kachkaev) [reply lib/markdownlint.js:911]: I shared some thoughts on `isFile()` in https://github.com/DavidAnson/markdownlint/pull/342#discussion_r507223896. At the moment, the only resulting error code is `ENOENT`, which can be emitted later from `readFile` / `readFileSync`. 

The current idea is that the resolve function never throws (as before). We could change this, but this would require more design work and could possibly make the change more breaking for the existing users / tools.

File: .github/workflows/nodejs.yml
Comment (kachkaev) [reply .github/workflows/nodejs.yml:33]: # 😱

File: test/markdownlint-test.js
Comment (kachkaev) [reply test/markdownlint-test.js]: Although there exists a [`test`](https://www.npmjs.com/package/test) package on npm, I renamed as you suggested. I don't think this will ever cause a clash.

Comment (kachkaev) [reply test/markdownlint-test.js]: Symlink creation should work on Windows too. CI does not seem to complain.

File: test/markdownlint-test.js
Comment (kachkaev) [reply test/markdownlint-test.js]: Symlink creation should work on Windows too. CI does not seem to complain.

File: lib/markdownlint.js
Comment (DavidAnson) [inline lib/markdownlint.js:904]: Nit: End with ".", please

File: lib/markdownlint.js
Comment (DavidAnson) [reply lib/markdownlint.js:911]: Ok

File: lib/markdownlint.js
Comment (DavidAnson) [reply lib/markdownlint.js:914]: Ok

File: lib/markdownlint.js
Comment (DavidAnson) [reply lib/markdownlint.js:923]: What you have here seems good! I'd prefer to get rid of the ESLint suppression's, I think the `error` parameter is unnecessary?

File: lib/markdownlint.js
Comment (DavidAnson) [inline lib/markdownlint.js:926]: Nit: I think in the rest of the file there are not blank lines between sections like this.

File: test/markdownlint-test.js
Comment (DavidAnson) [reply test/markdownlint-test.js:1099]: I was not meaning to suggest adding special error handling, just that if the new behavior gives a better error, that is nice!

File: test/markdownlint-test.js
Comment (DavidAnson) [inline test/markdownlint-test.js:1041]: I think I wasn't clear. Instead of adding to root/node_modules, can we add to root/test/node_modules? That has the benefit of not touching the "real" one and also showing up in git status if something goes wrong.

File: test/markdownlint-test.js
Comment (DavidAnson) [reply test/markdownlint-test.js:1102]: Agreed. But this test is named "...package", so I thought its error messages should say "package" instead of "file".

**Review by DavidAnson — CHANGES_REQUESTED** [review]
Thanks again! Left a couple more minor comments. This is probably the last round since I am getting nit-picky. I may end up tweaking the changes very slightly after merging, but that's better than giving you a hard time.

**Comment by DavidAnson** [comment]
Please also update the PR to be against next branch instead of main.

File: lib/markdownlint.js
Comment (kachkaev) [reply lib/markdownlint.js:923]: I tried a few things here, but could not get rid of an ESLint error. Having `} catch {` instead of `} catch (error) {` produces:

```
Parsing error: Unexpected token { eslint
```

Adding it back gives:

```
Remove unused catch binding `error`. eslint(unicorn/prefer-optional-catch-binding)
```

(which is what I am suppressing the same way as this is done in a few existing cases)

File: test/markdownlint-test.js
Comment (kachkaev) [reply test/markdownlint-test.js]: IC, IC, yeah I misunderstood what you were suggesting. I moved the symlink to `root/test/node_modules` now. What we can do as well is just to commit `root/test/node_modules/pseudo-package` instead and thus avid the symlinking complexity. This, however, whould require some change in `.gitignore` so I decided to not do this without your pre-approval.

**Comment by kachkaev** [comment]
Thanks for another review @DavidAnson! Happy to do address more comments if you want – no rush with the merge. Let’s do it properly 💪

File: test/markdownlint-test.js
Comment (DavidAnson) [inline test/markdownlint-test.js:1047]: Could we have the directory name match this test name? One day, maybe there will be another.

**Review by DavidAnson — APPROVED** [review]
Looks great! I left one more tiny comment. I'll probably release the patch version in the next couple of days, then can merge this into `next` branch.

Thanks!!

**Comment by fisker** [comment]
I'm the author of `unicorn/prefer-optional-catch-binding`, why we enabled this rule, but used a lot `// eslint-disable-next-line unicorn/prefer-optional-catch-binding`?

**Comment by DavidAnson** [comment]
@fisker That rule is enabled because I agree with it and it looks to be supported by Node version 10+ (as this library is). You may know better, but I'm guessing that I have ESLint slightly misconfigured to be using a language version that does not allow the optional binding? I will look into that and fix it if so.

**Comment by fisker** [comment]
Ah, the `ecmaVersion` is misconfiged, https://github.com/DavidAnson/markdownlint/blob/82cf68023f7dbd2948a65c53fc30482432195de4/.eslintrc.json#L3

I think it should be at least `2019`?

**Comment by DavidAnson** [comment]
@fisker Yes, probably. If there is a good mapping from Node version to this setting, I'd live to know! I will update this tomorrow and remove the unnecessary suppressions. Thanks!

**Comment by fisker** [comment]
This? https://node.green/#ES2019

**Comment by DavidAnson** [comment]
@fisker I like that site and it's what I use, but note that  some of those features are supported by Node 10 and some are not. I guess what I really want is to specify the minimum Node version to ESLint rather than the JavaScript version. No big deal; I will go with 2019 and make the associated updates. Thanks again!

File: test/markdownlint-test.js
Comment (kachkaev) [reply test/markdownlint-test.js:1047]: Do you mean renaming `./config/pseudo-package`, `test` or both? Happy to do so, just need a bit more detail. The test is called `configMultipleWithRequireResolve` (camel-case), which we can’t reuse in dir names without changing.

Comment (DavidAnson) [reply test/markdownlint-test.js:1047]: Sorry, I was thinking something like:  fakeNodeModule/configMultipleWithRequireResolve

Comment (kachkaev) [reply test/markdownlint-test.js:1047]: `test/configs/fakeNodeModule/configMultipleWithRequireResolve/package.json` plus `test/configs/fakeNodeModule/configMultipleWithRequireResolve/config.json` ?

or

`test/configs/fakeNodeModule/package.json` plus `test/configs/fakeNodeModule/configMultipleWithRequireResolve.json`?

Sorry got confused a bit. 

[If we look at Prettier](https://github.com/prettier/prettier/tree/master/tests_integration/plugins), they just store `node_modules/whatever` right in their repo, so that they don't need to symlink the files. We could to something similar by moving `test/configs/pseudo-package` to `test/node_modules/pseudo-package` or whatever name that would closer match the test name. I would not use `configMultipleWithRequireResolve` because it's camelCase, which is [not allowed for npm package names](https://docs.npmjs.com/files/package.json#name).

Comment (kachkaev) [reply test/markdownlint-test.js:1047]: Happy for you to push into this PR if that's less time-consuming 🙂

Comment (DavidAnson) [reply test/markdownlint-test.js:1047]: test/node_modules/pseudo-package sounds great! Agree with adding an exclude line to .gitignore for test/node_modules. Or neither and let me tweak a bit when I come back to this after the patch release. 😀

Comment (kachkaev) [reply test/markdownlint-test.js]: Done! 👍

Comment (DavidAnson) [reply test/markdownlint-test.js]: Love it! No copy, no clean-up.

**Comment by kachkaev** [comment]
Thanks again for reviewing and merging this PR @DavidAnson! Looking forward to trying it out in the released version 🚀

**Comment by kachkaev** [comment]
👋 @DavidAnson! What are your thoughts on releasing the next version, which would include this change?

**Comment by DavidAnson** [comment]
I picked out 6 "enhancement" issues to include with the next release. I'm about half way through them now. Once done, I plan to publish!

**Pull request merged**
